lib/ecto/repo/model.ex

```diff
@@ -176,6 +176,38 @@ defmodule Ecto.Repo.Model do
   end
 
   @doc """
+  Implementation for `Ecto.Repo.insert_or_update/2`.
+  """
+  def insert_or_update(repo, adapter, changeset, opts) do
+    case get_state(changeset) do
+      :built  -> insert repo, adapter, changeset, opts
+      :loaded -> update repo, adapter, changeset, opts
+      state   -> raise ArgumentError, "the changeset has an invalid state " <>
+                                      "for Repo.insert_or_update/2: #{state}"
+    end
+  end
+
+  @doc """
+  Implementation for `Ecto.Repo.insert_or_update!/2`.
+  """
+  def insert_or_update!(repo, adapter, changeset, opts) do
+    case get_state(changeset) do
+      :built  -> insert! repo, adapter, changeset, opts
+      :loaded -> update! repo, adapter, changeset, opts
+      state   -> raise ArgumentError, "the changeset has an invalid state " <>
+                                      "for Repo.insert_or_update/2: #{state}"
+    end
+  end
+
+  defp get_state(%Changeset{model: model}), do: model.__meta__.state
+  defp get_State(%{__struct__: model}) do
```

Comment:
Thanks for seeing that @brianstorti - https://github.com/elixir-lang/ecto/pull/1063